### PR TITLE
[feat] Fix Vllm, logprob, add filtering, temperature annealing, lr descent

### PR DESCRIPTION
**Pull request by YeAnbang**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ColossalChat/coati/distributed/consumer.py [hunk 1]
```diff
@@ -121,6 +122,8 @@ def loop(self) -> None:
                                 pbar.set_postfix({"loss": loss})
                             i += 1
                     assert len(self.buffer) == 0
+                    if self.lr_scheduler is not None:
+                        self.lr_scheduler.step()
```
Comment (TongLi3701) [inline applications/ColossalChat/coati/distributed/consumer.py:126]: why do we update lr scheduler here? 

The lr should be updated every step for policy update so I assume it should be inside the step function?

File: applications/ColossalChat/coati/distributed/inference_backend.py [hunk 2]
```diff
@@ -175,27 +187,40 @@ class VLLMInferenceBackend(BaseInferenceBackend):
...
     def generate(self, input_ids: torch.Tensor, attention_mask: torch.Tensor, **kwargs) -> Dict[str, torch.Tensor]:
         micro_batch_size = input_ids.size(0)
         response_start_idx = input_ids.size(1)
+        micro_batch_input_ids = input_ids.tolist()
+        micro_batch_input_ids_no_padding = []
+        for i in range(micro_batch_size):
+            for j in range(input_ids.size(1)):
+                if micro_batch_input_ids[i][j] != self.tokenizer.pad_token_id:
+                    micro_batch_input_ids_no_padding.append(micro_batch_input_ids[i][j:])
+                    break
```
Comment (TongLi3701) [inline applications/ColossalChat/coati/distributed/inference_backend.py:221]: Can be optimized to avoid nested for loop?

Comment (TongLi3701) [reply applications/ColossalChat/coati/distributed/inference_backend.py:221]:
```  micro_batch_input_ids_no_padding = []
        for i in range(micro_batch_size):
            # 找到第一个非填充标记的索引
            non_pad_index = (input_ids[i] != self.tokenizer.pad_token_id).nonzero(as_tuple=True)[0]
            if non_pad_index.numel() > 0:
                first_non_pad_index = non_pad_index[0].item()
                # 截取非填充部分
                non_padded = input_ids[i, first_non_pad_index:].tolist()
                micro_batch_input_ids_no_padding.append(non_padded)
        return micro_batch_input_ids_no_padding

File: applications/ColossalChat/coati/distributed/producer.py [hunk 3]
```diff
@@ -117,6 +117,12 @@ def loop(self) -> None:
                         None, self.num_producers, device=self.device, group_name="sync_model"
                     )
                     self.load_state_dict(state_dict)
+                # linear annealing for 1 episode, temperature from initial to 0.7
+                if episode <= 0:
+                    ratio = 1 - (len(self.dataloader) - i) / len(self.dataloader)
+                    self.model.generate_config.temperature = (
+                        ratio * self.generate_config["temperature"] + (1 - ratio) * 0.7
+                    )
```
Comment (TongLi3701) [inline applications/ColossalChat/coati/distributed/producer.py:125]: I assume here is the same? we should count step from the training side?

**Review by TongLi3701 — COMMENTED** [review]
Thanks Anbang, overall looks good.

I just have a few comments, please have a look.

File: applications/ColossalChat/coati/distributed/consumer.py (hunk 1 above)
Comment (TongLi3701) [reply applications/ColossalChat/coati/distributed/consumer.py:126]: In addition, we need to log LR and temperature as well for checking purpose.

Comment (YeAnbang) [reply applications/ColossalChat/coati/distributed/consumer.py:126]: for lr scheduler, I think updating lr per step make no difference to updating lr per producer step, and the later is easier to implement while the former need to consider a lot of facts in order to calculate the total number of consumer steps.

File: applications/ColossalChat/coati/distributed/producer.py (hunk 3 above)
Comment (YeAnbang) [reply applications/ColossalChat/coati/distributed/producer.py:128]: the consumer and producer are two isolated ray processes and hence cannot do this

**Review by TongLi3701 — APPROVED** [review]
Thanks Anbang, 

Overall look good.
